feat: add wrapper function for low level calls

julianmrodri opened the pull request:
<!-- 0. 🎉 Thank you for submitting a PR! -->

<!-- 1. Does this close any open issues? Please list them below. -->
#2260 
<!-- Keep in mind that new features have a better chance of being merged fast if
they were first discussed and designed with the maintainers. If there is no
corresponding issue, please consider opening one for discussion first! -->

Fixes #2260 

<!-- 2. Describe the changes introduced in this pull request. -->

This PR creates a function `Address.functionCall` which can be used to wrap low level calls that need to be performed on other Smart contracts. .

It also modifies the `ERC721.sol` and `SafeERC20.sol`contracts to use this new function instead of performing directly the low level call. Tests for these contracts needed to be updated to match the new revert messages in certain situations.


<!-- 3. Before submitting, please make sure that you have:
  - reviewed the OpenZeppelin Contributor Guidelines
    (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/CONTRIBUTING.md),
  - added tests where applicable to test new functionality,
  - made sure that your contracts are well-documented,
  - run the Solidity linter (`npm run lint:sol`) and fixed any issues,
  - run the JS linter and fixed any issues (`npm run lint:fix`), and
  - updated the changelog, if applicable.
-->

frangio commented on contracts/utils/Address.sol:
```diff
@@ -57,4 +57,31 @@ library Address {
+                    revert(add(32, returndata), returndata_size)
+                }
+            } else {
+                revert("Address: low-level call failed");
```
What we can do is receive this string as an argument to `functionCall`, and have an override without the third argument that uses this default reason.

nventuro replied:
Yes, this is what I was thinking. This would be the same approach as in [SafeMath.sub](https://docs.openzeppelin.com/contracts/3.x/api/math#SafeMath-sub-uint256-uint256-string-).

julianmrodri replied:
Great. I'll implement that.

frangio commented:
This is looking great!

julianmrodri commented:
Thanks!  I think I might need some guidance on the Tests to include. Which kind of `calls` should I mock?  Do you have any preference on what contract to use and what samples to use for the Unit tests?

nventuro commented:
We might want to also add a `value` field for ETH transfers.

>Thanks! I think I might need some guidance on the Tests to include. 

Off the top of my head, interesting scenarios are the following (thought we might not need to test them all):

- call to a contract that implements the called function and returns some data
- call to a contract that implements the called function but reverts
- call to a contract that implements the called function but runs out of gas
- call to a contract that doesn't implement the function (and reverts)
- call to a non-contract

If we add a `value` field, we might also want to test functions that are not `payable`.

julianmrodri commented:
@nventuro @frangio I've added the error message parameter so now we have the two options for `functionCall` (with or without the custom message). I also added some of the tests. 

I would like you to review the TESTS to see if this is the correct direction. I had to make too many decisions and is better for you to validate if this is OK. Feel free to let me know if this is not the correct way of doing this!

So far I covered:

* Call to a contract that implements the called function successfully
* Call to a contract that implements the called function but reverts
* Call to a contract that doesn't implement the function (and reverts)
* Call to a non-contract

I was **not** able to test the case of a function that returns data with this approach. It is quite challenging as I get the **Transaction** info and not the **returned value**, to eventually decode. Do you know how this can be achieved? Thanks!

nventuro commented on contracts/utils/Address.sol:
```diff
@@ -57,4 +57,42 @@ library Address {
+
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory returndata) = target.call(data);
+        if (!success) {
```
Can we have this be `if (success)` and swap the if and else blocks? I find the code more readable that way.

julianmrodri replied:
Yes I dont see why not. I took that from the original implementation but we can swap.

nventuro commented on test/token/ERC20/SafeERC20.test.js:
```diff
@@ -15,7 +15,7 @@ describe('SafeERC20', function () {
     });
 
-    shouldRevertOnAllCalls('SafeERC20: call to non-contract');
+    shouldRevertOnAllCalls('Address: call to non-contract');
```
@frangio are we ok with changing revert reasons in non-major releases?

(I think we are)

frangio replied:
Hm not possible to do that kind of thing dynamically. I don't think revert reasons should be stable, it's too limiting for us. I'm not 100% sure, but I think it should be ok to change it.

nventuro replied:
If we _really_ wanted to keep the old message, we could have an extra `Address.isContract()` check on `SafeERC20` and have that fail first, but I agree with @frangio here and we can leave as-is.

nventuro commented on contracts/mocks/AddressImpl.sol:
```diff
@@ -13,6 +13,10 @@ contract AddressImpl {
     }
 
+    function functionCall(address target, bytes calldata data) external returns (bytes memory) {
+        return Address.functionCall(target, data);
```
Because this function is not `view`, accessing the returned data is not possible. The easiest way to test this is to emit an event including it. See for example the tests for `EnumerableSet`.

nventuro replied:
@julianmrodri this is what you were asking about

julianmrodri replied:
Thanks @nventuro will look into this. The rest of the tests are kind of OK-ish?

frangio commented on contracts/utils/Address.sol:
```diff
@@ -57,4 +57,42 @@ library Address {
+     * call with `data` to the target address `target`. Returns the `returndata`
+     * provided by the low-level call. Uses `errorMessage` as default revert message.
+     
+     * The call is not executed if the target address is not a contract.
```
The calls should also mention that the function will bubble up the revert if the function call reverts.

frangio commented on contracts/utils/Address.sol:
```diff
@@ -57,4 +57,42 @@ library Address {
     }
+
+    /**
+     * @dev Replacement for Solidity's low-level `call`: performs a low-level
```
Instead of repeating the same docs here, we would put a link to the other one ("See `functionCall`") and only explain what's different about this one (in this case the default revert reason).

I'm not sure how we can automatically generate a link to a function with an overload like this one. @nventuro do you remember?

nventuro replied:
What I do is just run the docsite manually and copy-paste the identifier. I can add this myself on top of this PR later on.

julianmrodri replied:
Yes. I repeated the docs because I saw it was done a similar approach in the [div](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/math/SafeMath.sol#L91-L125) example. But I agree better not to repeat.

nventuro replied:
Ah that's ineteresting, thanks. We should improve those.

nventuro commented:
By the way I [asked the Solidity dev team on gitter](https://gitter.im/ethereum/solidity-dev?at=5edfb465013105125a245b86) if they think our checks are correct, and they gave us green light:

>>>
nventuro
hello! I am trying to replicate Solidity function call semantics using call, and wanted to double-check I'm not missing anything. This is what I'm currently doing:

    check using extcodehash that the recipient account has code, revert otherwise
    perform the call
    if success, abi.decode the returndata
    if failure, check the returndata for a revert reason (lenght > 0). if available, revert with that, otherwise revert with a generic reason

Is that all?

chriseth
@nventuro looks good!

nventuro commented:
@julianmrodri no need to apologize! Your help is greatly appreciated :)

nventuro commented:
For some reason the links are not working (see https://deploy-preview-2264--openzeppelin-contracts-docs.netlify.app/contracts/3.x/api/utils#Address-functionCall-address-bytes-string-).

julianmrodri commented:
Wow you took this to the next level. Awesome!

frangio commented on contracts/utils/Address.sol:
```diff
@@ -59,30 +59,69 @@ library Address {
-     
-     * The call is not executed if the target address is not a contract.
+     * @dev Same as {Address-functionCall-address-bytes-}, but with
+     * `errorMessage` as a custom revert reason when `target` reverts.
```
I think this conveys the idea a bit better.

```suggestion
     * `errorMessage` as a fallback revert reason when `target` reverts.
```

frangio commented on contracts/utils/Address.sol:
```diff
@@ -59,30 +59,69 @@ library Address {
+
+    /**
+     * @dev Same as {Address-functionCallWithValue-address-bytes-uint256-}, but
+     * with `errorMessage` as a custom revert reason when `target` reverts.
```
```suggestion
     * with `errorMessage` as a fallback revert reason when `target` reverts.
```

frangio commented on contracts/utils/Address.sol:
```diff
@@ -59,30 +59,69 @@ library Address {
+     * with `errorMessage` as a custom revert reason when `target` reverts.
+     */
+    function functionCallWithValue(address target, bytes memory data, uint256 value, string memory errorMessage) internal returns (bytes memory) {
+        require(address(this).balance >= value, "Address: insufficient balance for call");
```
This is interesting, does Solidity do this check too?

nventuro replied:
No, the `call` reverts, so `success` is false, and you get the generic `low level call failed` revert reason. I added this check since that balance check is extremely cheap: https://eips.ethereum.org/EIPS/eip-1884#specification

frangio replied:
I wouldn't call that "extremely cheap"!! In fact, it's the same as the base cost of `CALL`, which should the dominating cost of `functionCall`. Adding this check increases that 2x. So I have a strong preference towards removing it.

frangio replied:
Hm, okay when transferring value the cost of a call increases by 9000 gas, so it would not be a 2x increase, but I'd still call it significant.

In any case, I think we should remove the check so as to be closer to Solidity semantics.

nventuro replied:
`balance` costs 700 gas, but `selfbalance` (what that expression compiles into) costs 5 gas. That's what I was linking, `selfbalance` was introduced in that EIP. Sorry if this caused confusion.

frangio replied:
Ah, my bad!

frangio commented on test/utils/Address.test.js:
```diff
@@ -145,4 +189,86 @@ describe('Address', function () {
+        );
+      });
+
+      it('calls the requested function with existing value', async function () {
```
We're not asserting that the target contract received the requested value.

nventuro replied:
Great catch, thanks!

frangio commented on test/utils/Address.test.js:
```diff
@@ -233,9 +233,13 @@ describe('Address', function () {
           inputs: [],
         }, []);
 
+        const tracker = await balance.tracker(this.contractRecipient.address);
```
:heart_eyes:

julianmrodri commented:
Congrats guys! Awesome job. Glad to be able to be part of this. 👍

nventuro commented:
Thanks a lot to you both!
